internal/types/document_validation.go

```diff
@@ -138,10 +140,24 @@ func (d *Document) ValidateData() error {
 			}
 		case float64:
 			if math.IsInf(v, 0) {
+				if isUpdate {
+					return newValidationError(
+						ErrValidation, fmt.Errorf("update produces invalid value: { %q: %f } "+
+							"(update operations that produce infinity values are not allowed)", key, v,
+						),
+					)
+				}
+
 				return newValidationError(
 					ErrValidation, fmt.Errorf("invalid value: { %q: %f } (infinity values are not allowed)", key, v),
 				)
 			}
+
+			if v == 0 && math.Signbit(v) {
+				return newValidationError(
+					ErrValidation, fmt.Errorf("update produces invalid value: { %q: -0 } (-0 is not supported)", key),
```

Comment:
diff.md it says "update operation producing -0 is converted to 0." Is there some place that process that error and returns 0 if it happens?

Reply:
Yeah diff.md is correct. However any future operator which may produce -0 by update operation may silently put the -0 in the DB. Having this check will avoid that.

Reply:
Other way of safety check is also possible here, do you have other suggestion? :hugs:

Reply:
I don't really understand how it works. I expect this error is checked somewhere, and `0` is returned instead of `-0`, but I don't see where.

Reply:
Before this change, update operation result `-0` was stored it to DB for the edge case that was not caught. Insert of `-0` is validated here so we can't insert it https://github.com/FerretDB/FerretDB/blob/main/internal/wire/validation.go#L74.

A document which may accidentally produce `-0` could be caught here and return error.

Reply:
Let's remove this validation and sets `-1` to `0`. I initially added this change because we sets -0 to 0 in $mul operator, but for other existing operators, edge case of inserting -0 might exist. And if it did update -0, it panic'd in https://github.com/FerretDB/FerretDB/pull/1760#discussion_r1068965770 which was undesirable.